[IRGen/Runtime] Update Function Metadata

xedin opened the pull request:
Series of changes to streamline the way function metadata is stored.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin commented:
@DougGregor could you please take a look at `visitFunctionType` changes I have so far?

slavapestov commented:
This looks cleaner. Is there a change of behavior here? I notice we no longer treat the argument list as a single tuple when its materializable.

xedin commented:
@slavapestov yes, I still need to take a look at `getFunctionTypeMetadata` @rjmccall suggested to see if I need to make any changes and write some tests which use coercion to verify that everything behaves correctly.

xedin commented:
Did some more digging tonight and it looks like when we try to convert function types into `SILFunctionType` that looses parens in `AbstractionPattern::getFunctionInputType()`, so even with changes to metadata I made, casts with and without parens still succeed because they point to the same function type...

DougGregor reviewed (commented):
This looks great to me. There are a couple of additional things we'll need to describe in metadata (e.g., `__shared`, argument labels), but those can follow.

DougGregor commented:
Needs tests to commit, of course ;)

xedin committed:
[IRGen] Update Function Metadata to use AnyFunctionType::Param

This changes layout of the parameter metadata from single tuple record
(in case of materializable type) to N records each corresponding to
invididual function parameter, where functions with no parameters
`() -> Void` get 0 records allocated.

xedin commented:
@DougGregor since my changes to mangling went in, I've updated this changes with test cases to verify that casts are executed correctly now as well as calls to metadata are emitted to the right `swift_getFunctionTypeMetadata` function. Please take a look.

xedin commented:
@swift-ci please clean smoke test

xedin commented:
Next up going to be adding labels and missing flags for each parameter metadata entry

xedin commented:
Ok, it looks like empty parameters have to be handles specially here as well as case `(()) -> Void` where it's actually 1 parameter of empty type. Will handle it tomorrow.

xedin commented:
@swift-ci please clean smoke test

DougGregor commented on lib/RemoteAST/RemoteAST.cpp:
```diff
@@ -342,6 +342,11 @@ class RemoteASTTypeBuilder {
+      if (elts.empty())
+        input = TupleType::getEmpty(Ctx);
+      else
+        input = TupleType::get(elts, Ctx);
```
You don't need this "if" here; `TupleType::get(elts, Ctx)` works with zero elements just as well.

xedin replied:
Ah, should have looked at what `getEmpty(...)` does, I though the fact we have it might produce something more optimized from memory point of view :)

DougGregor replied:
It'll hit hash table and return the same thing as `TupleType::getEmpty(Ctx)`.  If there's anything to optimize here, it's an `empty()` check in `TupleType::get()` itself.

DougGregor replied:
It doesn't really matter either way.

DougGregor commented on test/IRGen/dynamic_cast_functions.swift:
```diff
@@ -0,0 +1,44 @@
+
+// CHECK: ok
+print((cast7 == nil) ? "ok" : "fail")
+// CHECK: ok
```
I think `assert` (or StdlibUnitTest) checks are preferred to FileCheck'ing the output nowadays. Is that right, @airspeedswift  ?

xedin replied:
That's what I found tests in `test/Interpreter/` do for dynamic casts...

DougGregor replied:
Yeah, it's fine. Ignore my comment :)

DougGregor reviewed (approved):
This looks fantastic. Thank you!

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please clean test
